### PR TITLE
auto generate msvc pch source file if none is provided by the user

### PR DESCRIPTION
**Pull request by nioncode**
As mentioned in #4682, meson can now auto generate the PCH implementation (which is only needed for msvc) if the user did not supply one.
Also, a warning will be printed if users supply a PCH implementation file to let them know that they should switch to single header PCH file.

### CODE REVIEW

**Comment by nioncode** [comment]
Test case fails on vc2015x86ninja with `c1: fatal error C1033: cannot open program database`. This seems to be some kind of concurrency issue, since it works fine on all other machines. Did something like this happen in the past already?

**Comment by nioncode** [comment]
Also, it looks like meson generates the project 3 times and compiles it twice. Might this be the cause for the error later on?

**Comment by nioncode** [comment]
Hmm I just repushed to see if this was just a CI hickup, but it fails again with the same error message only on vc2015x86ninja.

@jpakkane do you know if the vc2015x86ninja machine is set up differently? I read about a ninja timestamp bug in some other meson issue, could this be related? It looks like the first build (with 22 steps) works fine and then somehow meson setup runs again, but then only has 18 steps and fails, because of some pdb access issue.

**Comment by nioncode** [comment]
I pushed a fix now to not update the created implementation file in case it already has the correct content. That way, ninja will not build the dependencies again and the issue might be fixed.

However, we should still investigate why the project gets regenerated at all, this seems to be severly broken on the CI, since I cannot reproduce this locally and the other CI machines also work fine.

**Comment by nioncode** [comment]
Well, now it fails for vc2017x64vs because the generated pch file does not exist, which is .. also weird. Any suggestions how to debug this?

**Comment by jpakkane** [comment]
1 : fatal error C1083: Cannot open source file: '2e7d2c0@@prog@exe\meson_pch-c.c': No such file or directory [D:\a\1\s\b c2ebc55647\c\2e7d2c0@@prog@exe.vcxproj]

Is that the correct path it should be looked up in? If yes, then maybe the generation has not been run yet or it wrote it to a wrong dir?

**Comment by nioncode** [comment]
Yes, it is the correct path and it works just fine for vs2015 on CI and I locally tested again with vs2017 and it also works. From the output you can see that generation has already run, so it is quite weird that the file is apparently not there.

Is there a way to see the workspace of the build to check if the file is really not there or maybe in a different place? I'll add a few more logging statements to debug this.

**Comment by nioncode** [comment]
Now it works somehow. Is this some kind of race condition? The only way is to run the CI a few times to see if it works or maybe merge it to master to have it run for all other CI runs and see if it breaks stuff.

**Comment by nioncode** [comment]
Ok, now it failed again and the debug prints show the issue. When creating the pch file, I use a relative path and the current working directory seems to sometimes not being set to the build directory, but most of the time it actually is.

@jpakkane Are there any guarantees by meson that the cwd is set to a specific directory during generation? Otherwise, I'll just use an absolute path with the build dir prepended, instead.

**Comment by nioncode** [comment]
Looks fine now. I squashed and rebased on master to give it the final CI test.

**Comment by nioncode** [comment]
@jpakkane CI looks fine now (failures are simply due to Azure being broken now). Ready from my side.

**Comment by jpakkane** [comment]
Trunk has been fixed. Please rebase so CI is run.

**Comment by nioncode** [comment]
Done.

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -515,6 +515,33 @@ def get_pch_include_args(self, compiler, target):
...
+        # not defined. Otherwise, we might create the file in the wrong path.
+        pch_file = os.path.join(self.build_dir, pch_rel_to_build)
+        os.makedirs(os.path.dirname(pch_file), exist_ok=True)
+
+        content = '#include "%s"' % os.path.basename(pch_header)
+        self.write_file_if_different(pch_file, content)
+        return pch_rel_to_build
+
+    @staticmethod
+    def write_file_if_different(file, content):
```
Comment (jpakkane) [inline mesonbuild/backend/backends.py]: There is already mesonlib.replace_if_different, use that instead.

**Comment by jpakkane** [comment]
It's hard to tell from the diff, but is at least one of the pch test cases set up so that it uses the pre-existing source file? There should be at least one test for that case as well.

File: mesonbuild/backend/backends.py (hunk 1 above)
Comment (nioncode) [reply mesonbuild/backend/backends.py]: Dang, I knew I saw something like this somewhere already. Done.

**Comment by nioncode** [comment]
Yes, there is a test in `userDefined` subdir. It will print a warning during configuration though, that implementation files are deprecated and one should just use a single header file instead. I see no reason why one would want to use an implementation file that contains more than `#include "pch.h"`, especially since this file will only be used on Windows.

**Comment by nioncode** [comment]
Now Azure CI seems to be broken again, oh well.

**Comment by jpakkane** [comment]
>  I see no reason why one would want to use an implementation file that contains more than `#include "pch.h"

I have said words to this effect several times. It turns out there is _always_ someone who has a reason and who will then complain.

**Comment by jpakkane** [comment]
Master should be fixed, so rebasing should produce all green.

**Comment by nioncode** [comment]
Shouldn't we wait for someone to complain about this first? It's only a warning, so things still work without issues.

**Comment by jpakkane** [comment]
Yes, I just wanted to make sure there was a test so we don't break it accidentally in the future.

**Pull request merged**